src/include/duckdb/common/arrow/appender/varchar_data.hpp

```diff
@@ -112,4 +112,76 @@ struct ArrowVarcharData {
 	}
 };
 
+struct ArrowVarcharToStringViewData {
+	static void Initialize(ArrowAppendData &result, const LogicalType &type, idx_t capacity) {
+		result.main_buffer.reserve((capacity + 1) * 16);
+		result.aux_buffer.reserve(capacity);
+	}
+
+	static void Append(ArrowAppendData &append_data, Vector &input, idx_t from, idx_t to, idx_t input_size) {
+		idx_t size = to - from;
+		UnifiedVectorFormat format;
+		input.ToUnifiedFormat(input_size, format);
+
+		// resize the validity mask and set up the validity buffer for iteration
+		ResizeValidity(append_data.validity, append_data.row_count + size);
+		auto validity_data = (uint8_t *)append_data.validity.data();
+
+		// resize the offset buffer - the offset buffer holds the offsets into the child array
+		auto data = UnifiedVectorFormat::GetData<string_t>(format);
+		idx_t last_offset = 0;
+		for (idx_t i = from; i < to; i++) {
+			auto source_idx = format.sel->get_index(i);
+
+			if (!format.validity.RowIsValid(source_idx)) {
+				// Null value
+				uint8_t current_bit;
+				idx_t current_byte;
+				GetBitPosition(append_data.row_count + i - from, current_byte, current_bit);
+				SetNull(append_data, validity_data, current_byte, current_bit);
+				continue;
+			}
+
+			// These two are now the same buffer
+			auto arrow_4byte = append_data.main_buffer.GetData<int32_t>();
+			auto string_data = append_data.main_buffer.GetData<char>();
+			// This is the 16 byte index
+			auto result_idx = append_data.row_count + i - from;
+			auto &string_length = arrow_4byte[result_idx * 4];
+			string_length = ArrowVarcharConverter::GetLength(data[source_idx]);
+
+			if (string_length <= 12) {
+				//	This string is inlined
+				//  | Bytes 0-3  | Bytes 4-15                            |
+				//  |------------|---------------------------------------|
+				//  | length     | data (padded with 0)                  |
+				memcpy(&string_data[i * 16 + 4], data[source_idx].GetData(), string_length);
```

Comment:
Nit: manually striding byte buffers like this seems worryingly brittle to me. Including the corresponding union would be more readable:

```c++
  union arrow_string_view_t {
    struct {
      int32_t length;
      uint8_t data[12];
    } inlined;

    struct {
      int32_t length;
      uint8_t prefix[4];
      int32_t buffer_index, offset;
    } ref;

    int32_t length() const { return inlined.length; }
    const uint8_t* inline_data() const { return is_inline() ? inlined.data : ref.prefix; }
  };
};
```